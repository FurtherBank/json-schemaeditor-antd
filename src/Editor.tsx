import React, { useEffect, useLayoutEffect, useMemo, useRef, useState } from "react"
import { createStore, Store } from "redux"
import { Provider } from "react-redux"
import Field from "./Field"
import { ajvInstance, reducer } from "./reducer"
import FieldDrawer from "./FieldDrawer"
import Alert from "antd/lib/alert"

import "./css/index.scss"

interface EditorProps {
  editionName: string
  onChange?: (data: any) => void | null
  data?: any
  schema: RootSchema | boolean
}

const EditorHook = (props: EditorProps) => {
  const { schema, data, onChange, editionName } = props
  const compileSchema = () => {
    let validate = undefined
    let schemaErrors = null
    try {
      validate = ajvInstance.compile(schema)
      return validate
    } catch (error) {
      schemaErrors = error
      return schemaErrors
    }
  }
  const change = () => {
    if (onChange && typeof onChange == "function") {
      onChange(store.getState().data)
    }
  }
  const initStore = () => {
    const initialState = {
      data: data,
      editionName: editionName,
      rootSchema: typeof validate === "function" ? schema : true,
      lastChangedRoute: [],
      lastChangedField: [],
      dataErrors: [],
      schemaErrors: typeof validate === "function" ? null : validate,
      cache: {
        ofCache: new Map(),
        propertyCache: new Map(),
        itemCache: new Map(),
      },
      validate: typeof validate === "function" ? validate : undefined,
    }
    const store = createStore(reducer, initialState)
    store.subscribe(change)
    return store
  }

  const validate = useMemo(compileSchema, [schema]) as Function | any
  const drawerRef = useRef(null) as React.RefObject<any>
  const store = initStore()

  const setDrawer = (...args: any[]) => {
    console.log("setDrawer", drawerRef.current)
    if (drawerRef.current) drawerRef.current.setDrawer(...args)
  }
  console.log("editor重新渲染")

  return (
    <Provider store={store}>
      {validate instanceof Function ? null : (
        <Alert message="Error" description={validate.toString()} type="error" showIcon />
      )}
      <Field route={[]} field={null} schemaEntry="#" setDrawer={setDrawer} />
      <FieldDrawer ref={drawerRef} />
    </Provider>
  )
}

export default React.memo(EditorHook)
