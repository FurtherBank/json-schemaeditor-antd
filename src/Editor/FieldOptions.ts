import _ from "lodash"
import { SchemaCache } from "."
import { FatherInfo, FieldProps } from "./Field"
import {
  absorbProperties,
  addRef,
  filterIter,
  filterObjSchema,
  findKeyRefs,
  getPathVal,
  getRefSchemaMap,
  getValueByPattern,
  jsonDataType,
} from "./utils"

const maxCollapseLayer = 3
const longFormats = ["row", "uri-reference"]
const extraLongFormats = ["multiline"]

export const gridOption = [
  { gutter: 0, column: 1 },
  { gutter: 0, column: 2, lg: 3, xl: 4, xxl: 5 },
  { gutter: 0, column: 4, lg: 6, xl: 8, xxl: 10 },
]
export const defaultTypeValue: any = {
  string: "",
  number: 0,
  integer: 0,
  object: {},
  array: [],
  null: null,
  boolean: false,
}

/**
 * 确定一个变量在enum中作为一个选项显示时的名称
 * @param v
 * @returns
 */
const toEnumName = (v: any) => {
  const t = jsonDataType(v)
  switch (t) {
  case "object":
    return v.hasOwnProperty("name") ? v.name.toString() : `Object[${Object.keys(v).length}]`
  case "array":
    return `Array[${v.length}]`
  case "null":
    return "null" // 注意 null 没有 toString
  default:
    return v.toString()
  }
}

/**
 * 确定一个schema在of中作为一个选项显示时的名称
 * @param schemaMap 模式映射
 * @returns
 */
const toOfName = (schemaMap: Map<string, boolean | Schema>) => {
  const schemas = filterObjSchema(schemaMap.values())
  const title = absorbProperties(schemas, "title")
  if (title) return title
  const type = absorbProperties(schemas, "type")
  if (type.length === 1) return type[0]
  return ""
}

/**
 * 确定schema是否可以短优化。条件：
 * 1. 类型确定且为string/number/bool/null，且没有oneof
 * 2. 有enum
 * @param schema
 */
const schemaShortable = (ref: string, rootSchema: RootSchema | boolean | undefined) => {
  const schemaMap = getRefSchemaMap(ref, rootSchema)

  if (schemaMap.has(ref) && schemaMap.get(ref) === false) return true

  const existToTrue = ["const", "enum"]
  for (let i = 0; i < existToTrue.length; i++) {
    const key = existToTrue[i]
    const keySchema = absorbProperties(schemaMap, key)
    if (keySchema !== undefined) return true
  }

  const types = absorbProperties(schemaMap, "type")
  if (types.length === 1) {
    const type = types[0]
    switch (type) {
    case "string":
      const format = absorbProperties(schemaMap, "format")
      if (format && (extraLongFormats.includes(format) || longFormats.includes(format))) return false
      return true
    case "number":
    case "integer":
    case "boolean":
    case "null":
      return true
    default:
      return false
    }
  }
  return false
}

export const getFormatType = (format: string | undefined) => {
  if (extraLongFormats.includes(format!)) return 2
  if (longFormats.includes(format!)) return 1
  return 0
}

/**
 * 当前json在既定的schema下是否可以创建新的属性。
 * @param props 
 * @param schemaCache
 */
const canSchemaCreate = (props: FieldProps, schemaCache: SchemaCache) => {
  const { data } = props
  const {valueSchemaMap} = schemaCache
  const dataType = jsonDataType(data)
  for (const schema of valueSchemaMap!.values()) {
    if (schema === false) return false
  }
  const filteredSchemas = filterObjSchema(valueSchemaMap!.values())
  let autoCompleteKeys: string[] = []
  switch (dataType) {
  case "object":
    /**
       * object可以创建新属性，需要关注的条件：
       * 1. patternProperties 不为空，我们默认 patternProperties 只要有可用的正则就肯定能再创建。
       * 2. additionalProperties 不为 false
       * 3. 不超过 maxLength
       */
    const nowKeys = Object.keys(data)
    return filteredSchemas.every((schema) => {
      const { maxProperties, properties, additionalProperties, patternProperties } = schema
      // 1. 长度验证
      if (maxProperties !== undefined && nowKeys.length >= maxProperties) return false
      // 在这之前先收集一下可以自动补全的字段。。。
      let restKeys = []
      if (properties) {
        restKeys = Object.keys(properties).filter((key) => !nowKeys.includes(key))
        autoCompleteKeys = autoCompleteKeys.concat(restKeys)
        return restKeys.length > 0
      }
      // 2. additionalProperties 验证
      if (additionalProperties !== false) return true
      // 3. patternProperties 有键
      if (patternProperties && Object.keys(patternProperties).length > 0) return true
      // 4. 有无剩余键
      return restKeys.length > 0
    })
      ? autoCompleteKeys
      : false
  case "array":
    return filteredSchemas.every((schema) => {
      const { maxItems, items, additionalItems } = schema
      const itemsLength = items instanceof Array && additionalItems === false ? items.length : +Infinity
      const maxLength = maxItems === undefined ? +Infinity : maxItems
      return maxLength < itemsLength ? data.length < maxLength : data.length < itemsLength
    })
      ? []
      : false
  default:
    return false
  }
}

/**
 * 判断该字段(来自一个对象)是否可以重新命名
 * @param props
 * @returns 返回字符串为不可命名，返回正则为命名范围，返回空串即可命名
 */
const canSchemaRename = (props: FieldProps, schemaCache: SchemaCache) => {
  const { field, schemaEntry, fatherInfo } = props
  const { entrySchemaMap, itemCache, propertyCache, valueSchemaMap } = schemaCache
  // 注意，一个模式的 title 看 entryMap，如果有of等不理他
  const title = absorbProperties(entrySchemaMap!, "title") as string | undefined

  if (field === null) {
    return title ? title : ''
  }
  // 不是根节点，保证 FatherInfo 一定存在。
  const { valueEntry: fatherValueEntry} = fatherInfo!
  const propertyCacheValue = fatherValueEntry ? propertyCache.get(fatherValueEntry) : null
  if (fatherInfo!.type === "array") {
    return title ? title + " " + field : field
  } else if (!propertyCacheValue) {
    return ""
  } else {
    const {props, patternProps} = propertyCacheValue

    if (props[field]) return title ? title : field

    const pattern = getValueByPattern(patternProps, field)
    if (pattern) return new RegExp(pattern)
    
    return ''
  }
}

/**
 * 判断 该字段是否可删除。可删除条件：
 * 1. `field === null` 意味着根字段，不可删除
 * 2. 如果父亲是数组，只要不在数组 items 里面即可删除
 * 3. 如果父亲是对象，只要不在 required 里面即可删除
 *
 * @param props
 * @param schemaCache
 */
const canDelete = (props: FieldProps, schemaCache: SchemaCache) => {
  const { fatherInfo, field } = props
  const { itemCache, propertyCache } = schemaCache

  if (field === null) return false
  if (fatherInfo) {
    const { valueEntry: fatherValueEntry } = fatherInfo
    switch (fatherInfo.type) {
    case "array":
      const itemCacheValue = fatherValueEntry ? itemCache.get(fatherValueEntry) : null
      const index = parseInt(field)
      if (itemCacheValue) {
        return itemCacheValue.itemLength ? index >= itemCacheValue.itemLength : true
      } else {
        return true
      }
    case "object":
      const propertyCacheValue = fatherValueEntry ? propertyCache.get(fatherValueEntry) : null
      if (propertyCacheValue) {
        return !propertyCacheValue.required.includes(field)
      } else {
        return true
      }
    default:
      console.log("意外的判断情况")
      return false
    }
  }
  return false
}

/**
 * 通过一个schemaEntry 得到schema，确定其创建时默认对象。
 * 允许找不到schema的场合，且前后变量保持最大兼容(未实装)
 * @param schemaCache
 * @param entry
 * @returns
 */
export const getDefaultValue = (schemaCache: SchemaCache, entry: string | undefined, nowData: any = undefined): any => {
  const { ofCache, itemCache, propertyCache, rootSchema } = schemaCache!
  if (!entry) return null
  const entryMap = getRefSchemaMap(entry, rootSchema)
  const nowDataType = jsonDataType(nowData)
  const propertyCacheValue = propertyCache.get(entry)
  // 0. 如果nowData是对象，就先剪掉不在列表中的属性，然后进行合并
  if (nowDataType === 'object' && propertyCacheValue) {
    const {props, patternProps} = propertyCacheValue
    for (const key of Object.keys(nowData)) {
      if (!props[key] && !getValueByPattern(patternProps, key)) delete nowData[key]
    }
  }
  // 1. 优先返回规定的 default 字段值(注意深拷贝，否则会形成对象环！)
  const defaultValue = absorbProperties(entryMap, "default")
  if (defaultValue !== undefined) {
    const defaultType = jsonDataType(defaultValue)
    if (defaultType === 'object' && nowDataType === 'object') {
      // 特殊：如果默认是 object，会采取最大合并
      return Object.assign({}, nowData, _.cloneDeep(defaultValue))
    }
    return _.cloneDeep(defaultValue)
  } else {
    // 2. 如果有 const/enum，采用其值
    const constValue = absorbProperties(entryMap, "const")
    if (constValue !== undefined) return _.cloneDeep(constValue)
    const enumValue = absorbProperties(entryMap, "enum")
    if (enumValue !== undefined) return _.cloneDeep(enumValue[0])
    // 3. oneOf/anyOf 选择第0项的schema返回
    const ofCacheValue = ofCache.get(entry)
    if (ofCacheValue) {
      return getDefaultValue(schemaCache, addRef(ofCacheValue.ofRef, "0")!)
    }
  }
  // 4. 按照 schema 寻找答案
  const allTypes = absorbProperties(entryMap, "type")
  if (allTypes.length > 0) {
    const type = allTypes[0]
    switch (type) {
    case "object":
      const result = jsonDataType(nowData) === 'object' ? _.clone(nowData) : {}
      const allPropsRef = findKeyRefs(entryMap, "properties", true, false) as string[]
      // const propertyCacheValue = propertyCache.get(entry)
      // todo: 裁剪不允许出现的属性

      for (const ref of allPropsRef) {
        // const properties = getPathVal(rootSchema, addRef(ref, 'properties')!)
        // 仅对 required 中的属性进行创建
        const required = getPathVal(rootSchema, addRef(ref, "required")!) || []
        for (const propName of required) {
          result[propName] = getDefaultValue(schemaCache, addRef(ref, "properties", propName))
        }
      }
      return result
    case "array":
      const itemsRef = findKeyRefs(entryMap, "items") as string | undefined
      const additionalItems = absorbProperties(entryMap, 'additionalItems')
      const itemCacheValue = itemCache.get(entry)

      const arrayResult = jsonDataType(nowData) === 'array' && additionalItems ? _.clone(nowData) : []

      if (itemCacheValue && itemCacheValue.itemLength !== undefined) {
        const { itemLength } = itemCacheValue
        for (let i = 0; i < itemLength; i++) {
          arrayResult[i] = getDefaultValue(schemaCache, addRef(itemsRef, i.toString()))
        }
      }
      return arrayResult
    default:
      break
    }
    return _.cloneDeep(defaultTypeValue[type])
  } else {
    return null
  }
}
export { maxCollapseLayer, toEnumName, toOfName, schemaShortable, canSchemaCreate, canSchemaRename, canDelete }
