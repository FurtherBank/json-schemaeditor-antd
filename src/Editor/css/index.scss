
.selectable-selectbox {
  z-index: 9000;
  position: absolute;
  cursor: default;
  background: none;
  border: 1px dashed grey;
}

.selectable-div, .selected-div {
  width: 180px;
  height: 28px;
  line-height: 28px;
  padding: 0 3px;
  line-height: normal;
  box-sizing: border-box;
  border: 1px dotted #bbbbbb;
  user-select:none;
}

.selectable-div {
  background-color: #ffffff;
  :hover {
    background-color: #cccccc;
  }
}

.selected-div {
  background-color: #99ffff;
  :hover {
    background-color: #6699ff;
  }
}