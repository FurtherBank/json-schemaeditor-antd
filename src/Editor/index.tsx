import React, { forwardRef, useCallback, useEffect, useImperativeHandle, useLayoutEffect, useMemo, useRef, useState } from "react"
import { createStore, Store } from "redux"
import { Provider } from "react-redux"
import Field from "./Field"
import { ajvInstance, Caches, itemSchemaCache, ofSchemaCache, propertySchemaCache, reducer } from "./reducer"
import FieldDrawer from "./FieldDrawer"
import Alert from "antd/lib/alert"
import { composeWithDevTools } from 'redux-devtools-extension'

import "./css/index.scss"
import _ from "lodash"

interface EditorProps {
  onChange?: (data: any) => void | null
  data?: any
  schema: RootSchema | true
}
export const CacheContext = React.createContext({
  ofCache: new Map(),
  propertyCache: new Map(),
  itemCache: new Map(),
  rootSchema: {}
} as ContextContent)

export interface ContextContent {
  ofCache: Map<string, ofSchemaCache | null>
  propertyCache: Map<string, propertySchemaCache | null>
  itemCache: Map<string, itemSchemaCache | null>
  rootSchema: RootSchema
}

export interface SchemaCache extends ContextContent {
  entrySchemaMap: Map<string, boolean | Schema>
  valueEntry: string | undefined
  valueSchemaMap: Map<string, boolean | Schema>
}

const emptyArray: never[] = []

const EditorHook = (props: EditorProps, ref: React.ForwardedRef<any>) => {
  const { schema, data, onChange } = props

  // 编译 schema，并通知 schema 已经改变。
  const compileSchema = () => {
    let validate = undefined
    let schemaErrors = null
    
    console.time('compile schema')
    try {
      validate = ajvInstance.compile(schema)
    } catch (error) {
      schemaErrors = error
    }
    console.timeEnd('compile schema')

    return validate ? validate : schemaErrors
  }
  
  const initStore = () => {
    const initialState = {
      data: data,
      dataErrors: [],
      validate: typeof validate === "function" ? validate : undefined,
    }

    const store = createStore(reducer, {
      past: [],
      present: initialState,
      future: []
    }, composeWithDevTools())

    const change = () => {
      const changedData = store.getState().present.data
      if (onChange && typeof onChange == "function") {
        onChange(changedData)
      }
    }
    store.subscribe(change)
    
    return store
  }

  const validate = useMemo(compileSchema, [schema]) as Function | any
  const store = useMemo(initStore, [schema])
  useImperativeHandle(ref, () => {return store}, [store])

  const caches = useMemo(() => {
    console.log('caches变化')
    return {
      ofCache: new Map(),
      propertyCache: new Map(),
      itemCache: new Map(),
      rootSchema: validate instanceof Function ? typeof schema !== 'boolean' ? schema : {} : {}
    }
  }, [schema])

  // 如果 data 更新来自外部，通过 setData 与 store 同步
  const presentData = store.getState().present.data
  if (data !== presentData) {
    console.log('检测到外部更新：', data, presentData)
    store.dispatch({
      type: 'setData',
      value: data
    })
  }

  // 详细抽屉功能
  const drawerRef = useRef(null) as React.RefObject<any>
  const setDrawer = useCallback((...args: any[]) => {
    console.log("setDrawer", drawerRef.current)
    if (drawerRef.current) drawerRef.current.setDrawer(...args)
  }, [drawerRef])

  return (
    <Provider store={store}>
      {validate instanceof Function ? null : (
        <Alert message="Error" description={validate.toString()} type="error" showIcon />
      )}
      <CacheContext.Provider value={caches}>
      <Field route={emptyArray} field={null} schemaEntry="#" setDrawer={setDrawer} />
      <FieldDrawer ref={drawerRef} />

      </CacheContext.Provider>
    </Provider>
  )
}

export default React.memo(forwardRef(EditorHook))
