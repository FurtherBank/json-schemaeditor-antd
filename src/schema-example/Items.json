{
    "$schema": "http://json-schema.org/draft-06/schema#",
    "type": "array",
    "items": {
        "anyOf": [
            {
                "$ref": "#/definitions/Item"
            },
            {
                "type": "null"
            }
        ]
    },
    "definitions": {
        "Item": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "id": {
                    "type": "integer"
                },
                "animationId": {
                    "type": "integer"
                },
                "consumable": {
                    "type": "boolean"
                },
                "damage": {
                    "$ref": "#/definitions/Damage"
                },
                "description": {
                    "type": "string"
                },
                "effects": {
                    "type": "array",
                    "items": {
                        "$ref": "#/definitions/Effect"
                    }
                },
                "hitType": {
                    "type": "integer"
                },
                "iconIndex": {
                    "type": "integer"
                },
                "itypeId": {
                    "type": "integer"
                },
                "name": {
                    "type": "string"
                },
                "note": {
                    "type": "string"
                },
                "occasion": {
                    "type": "integer"
                },
                "price": {
                    "type": "integer"
                },
                "repeats": {
                    "type": "integer"
                },
                "scope": {
                    "type": "integer"
                },
                "speed": {
                    "type": "integer"
                },
                "successRate": {
                    "type": "integer"
                },
                "tpGain": {
                    "type": "integer"
                }
            },
            "required": [
                "animationId",
                "consumable",
                "damage",
                "description",
                "effects",
                "hitType",
                "iconIndex",
                "id",
                "itypeId",
                "name",
                "note",
                "occasion",
                "price",
                "repeats",
                "scope",
                "speed",
                "successRate",
                "tpGain"
            ],
            "title": "Item"
        },
        "Damage": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "critical": {
                    "type": "boolean"
                },
                "elementId": {
                    "type": "integer"
                },
                "formula": {
                    "type": "string"
                },
                "type": {
                    "type": "integer"
                },
                "variance": {
                    "type": "integer"
                }
            },
            "required": [
                "critical",
                "elementId",
                "formula",
                "type",
                "variance"
            ],
            "title": "Damage"
        },
        "Effect": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "code": {
                    "type": "integer"
                },
                "dataId": {
                    "type": "integer"
                },
                "value1": {
                    "type": "number"
                },
                "value2": {
                    "type": "integer"
                }
            },
            "required": [
                "code",
                "dataId",
                "value1",
                "value2"
            ],
            "title": "Effect"
        }
    }
}