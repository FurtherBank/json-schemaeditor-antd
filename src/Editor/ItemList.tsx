import React, { memo, useState } from "react"
import DataItem from "./DataItem"
import { ChildData } from "./Field"
import { toEnumName } from "./FieldOptions"

type Props = {
  items: ChildData[]
}

const ItemList = memo((props: Props) => {
  const { items } = props

  return (
    <div style={{ height: "100vh", overflow: "auto" }}>
      {items.map((item, i) => {
        const { value } = item
        return <DataItem key={i} id={i}>{`${i}. ${toEnumName(value)}`}</DataItem>
      })}
    </div>
  )
})

export default ItemList
