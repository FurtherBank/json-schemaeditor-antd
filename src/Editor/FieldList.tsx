import { PlusOutlined } from "@ant-design/icons"
import { Layout } from "antd"
import Button from "antd/lib/button"
import { Content } from "antd/lib/layout/layout"
import Sider from "antd/lib/layout/Sider"
import List from "antd/lib/list"
import space from "antd/lib/space"
import React, { useState } from "react"
import { SelectableGroup } from "react-selectable-fast"
import { SchemaCache } from "."
import CreateName from "./CreateName"
import { DataItemProps } from "./DataItem"
import Field, { FatherInfo, FieldProps } from "./Field"
import { gridOption } from "./FieldOptions"
import ItemList from "./ItemList"
import { ShortOpt } from "./reducer"
import { concatAccess, getFieldSchema, jsonDataType } from "./utils"

interface FieldListProps {
  fieldProps: FieldProps
  fieldCache: SchemaCache
  short: ShortOpt
  content: any[]
  fatherInfo: FatherInfo
  canCreate?: boolean
  view?: string
}

const FieldList = (props: FieldListProps) => {
  const { content, fieldProps, fatherInfo, short, canCreate, view, fieldCache } = props
  const doAction = fieldProps.doAction!
  const { data, route, field, setDrawer } = fieldProps
  const dataType = jsonDataType(data)
  const access = concatAccess(route, field)

  // items 后处理
  const endItem = { end: true, key: "", value: "" }
  const items = canCreate ? content.concat(endItem) : content

  // 对数组json专用的 列表选择特性
  const [currentItem, setCurrentItem] = useState(0)

  const handleSelectable = (selectedItems: React.Component<DataItemProps>[]) => {
    const ids: number[] = selectedItems.map((v) => {
      return v.props.id
    })
    if (ids.length > 0) {
      setCurrentItem(ids[0])
    }
  }

  switch (view) {
    case "list":
      return (
        <Layout style={{ height: "100%", flexDirection: "row",  alignItems: 'stretch', display: "flex" }}>
          <Sider theme={"light"} style={{ height: "100%"}}>
            <div style={{ height: "100%", display: "flex", flexDirection: "column" }}>
              <SelectableGroup
                clickClassName="tick"
                enableDeselect={true}
                tolerance={0}
                deselectOnEsc={false}
                allowClickWithoutSelected={true}
                resetOnStart={true}
                onSelectionFinish={handleSelectable}
                ignoreList={[".not-selectable"]}
                style={{ flex: "1", overflow: "auto", margin:"3px 0"}}
                key={'select'}
              >
                <ItemList items={content} />
              </SelectableGroup>
              {canCreate ? <CreateName fatherInfo={fatherInfo} fieldProps={fieldProps} fieldCache={fieldCache} style={{margin: "3px", width:"auto"}} key={'create'}/> : null}
            </div>
          </Sider>
          <Content style={{ height: "100%", overflow: 'auto' }}>
            {data.length > 0 ?
              <Field
              route={access}
              field={currentItem.toString()}
              fatherInfo={fatherInfo.type ? fatherInfo : undefined}
              schemaEntry={getFieldSchema(fieldProps, fieldCache, currentItem.toString())}
              short={short}
              setDrawer={setDrawer}
            /> : null}
          </Content>
        </Layout>
      )
    default:
      return (
        <List
          size="small"
          split={false}
          dataSource={items}
          grid={gridOption[short]}
          pagination={
            items.length > 16
              ? {
                  simple: true,
                  pageSize: 16,
                }
              : undefined
          }
          renderItem={(item) => {
            const { key, end, data: itemData } = item
            if (itemData) {
              return (
                <FieldList
                  key="start"
                  content={itemData}
                  fieldProps={fieldProps}
                  fieldCache={fieldCache}
                  fatherInfo={fatherInfo}
                  short={ShortOpt.short}
                />
              )
            } else if (!end)
              return (
                <List.Item key={"property-" + key}>
                  <Field
                    route={access}
                    field={key}
                    fatherInfo={fatherInfo.type ? fatherInfo : undefined}
                    schemaEntry={getFieldSchema(fieldProps, fieldCache, key)}
                    short={short}
                    setDrawer={setDrawer}
                  />
                </List.Item>
              )
            else
              return (
                <List.Item key="end">
                  <CreateName fatherInfo={fatherInfo} fieldProps={fieldProps} fieldCache={fieldCache}/>
                </List.Item>
              )
          }}
        />
      )
  }
}

export default FieldList
