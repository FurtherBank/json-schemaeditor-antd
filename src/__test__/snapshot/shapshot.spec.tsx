import React from "react"
import { cleanup, render, screen } from "@testing-library/react"
import Editor from "../../Editor"

import $Items from "../../schema-example/Itemsv20.json"
import Items from "../../json-example/Items.json"

import meta from "../../json-example/$meta.json"
afterEach(cleanup)

test("itemSnapshot", () => {
  const { asFragment } = render(<Editor data={Items} schema={$Items} />)
  expect(asFragment()).toMatchSnapshot()
})
