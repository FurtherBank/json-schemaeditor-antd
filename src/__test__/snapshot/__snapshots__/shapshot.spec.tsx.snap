// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`itemSnapshot 1`] = `
<DocumentFragment>
  <section
    class="ant-layout ant-layout-has-sider"
    style="height: 100%; flex-direction: row; align-items: stretch; display: flex;"
  >
    <aside
      class="ant-layout-sider ant-layout-sider-light"
      style="height: 100%; flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
    >
      <div
        class="ant-layout-sider-children"
      >
        <div
          style="height: 100%; display: flex; flex-direction: column;"
        >
          <div
            class="undefined "
            style="position: relative; flex: 1; overflow: auto; margin: 3px 0px;"
          >
            <div
              style="height: 100%; overflow: auto;"
            >
              <div
                class="selectable-div"
              >
                0. null
              </div>
              <div
                class="selectable-div"
              >
                1. -----保留
              </div>
              <div
                class="selectable-div"
              >
                2. 
              </div>
              <div
                class="selectable-div"
              >
                3. 
              </div>
              <div
                class="selectable-div"
              >
                4. 
              </div>
              <div
                class="selectable-div"
              >
                5. 
              </div>
              <div
                class="selectable-div"
              >
                6. -----恢复物品
              </div>
              <div
                class="selectable-div"
              >
                7. 药水
              </div>
              <div
                class="selectable-div"
              >
                8. 超级药水
              </div>
              <div
                class="selectable-div"
              >
                9. 痊愈药水
              </div>
              <div
                class="selectable-div"
              >
                10. 魔力水
              </div>
              <div
                class="selectable-div"
              >
                11. 兴奋剂
              </div>
              <div
                class="selectable-div"
              >
                12. 驱散草药
              </div>
              <div
                class="selectable-div"
              >
                13. 解毒剂
              </div>
              <div
                class="selectable-div"
              >
                14. 奇迹水滴
              </div>
              <div
                class="selectable-div"
              >
                15. 万能药
              </div>
              <div
                class="selectable-div"
              >
                16. 
              </div>
              <div
                class="selectable-div"
              >
                17. -----特别物品
              </div>
              <div
                class="selectable-div"
              >
                18. 遇敌减少器
              </div>
              <div
                class="selectable-div"
              >
                19. 掉落增加器
              </div>
              <div
                class="selectable-div"
              >
                20. 
              </div>
              <div
                class="selectable-div"
              >
                21. 
              </div>
              <div
                class="selectable-div"
              >
                22. -----增强物品
              </div>
              <div
                class="selectable-div"
              >
                23. HP增强剂
              </div>
              <div
                class="selectable-div"
              >
                24. MP增强剂
              </div>
              <div
                class="selectable-div"
              >
                25. 攻击增强剂
              </div>
              <div
                class="selectable-div"
              >
                26. 防御增强剂
              </div>
              <div
                class="selectable-div"
              >
                27. 魔力增强剂
              </div>
              <div
                class="selectable-div"
              >
                28. 抗性增强剂
              </div>
              <div
                class="selectable-div"
              >
                29. 敏捷增强剂
              </div>
              <div
                class="selectable-div"
              >
                30. 幸运增强剂
              </div>
            </div>
            <div
              class="selectable-selectbox"
              style="left: 0px; top: 0px; width: 0px; height: 0px; z-index: 9000; position: absolute; cursor: default; will-change: transform; transform: translateZ(0);"
            />
          </div>
          <button
            class="ant-btn ant-btn-dashed ant-btn-sm ant-btn-block"
            style="margin: 3px; width: auto;"
            type="button"
          >
            <span
              aria-label="plus"
              class="anticon anticon-plus"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="plus"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <defs>
                  <style />
                </defs>
                <path
                  d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                />
                <path
                  d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                />
              </svg>
            </span>
            <span>
              Item
            </span>
          </button>
        </div>
      </div>
    </aside>
    <main
      class="ant-layout-content"
      style="height: 100%; overflow: auto;"
    >
      <div
        class="ant-card ant-card-bordered ant-card-small"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              <div
                class="ant-space ant-space-horizontal ant-space-align-center"
              >
                <div
                  class="ant-space-item"
                >
                  <span
                    style="width: 100px;"
                    title="0"
                  >
                    0
                  </span>
                </div>
              </div>
            </div>
            <div
              class="ant-card-extra"
            >
              <div
                class="ant-space ant-space-horizontal ant-space-align-center"
              >
                <div
                  class="ant-space-item"
                  style="margin-right: 8px;"
                >
                  <span>
                    null
                  </span>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn ant-btn-circle ant-btn-default ant-btn-sm ant-btn-icon-only"
                    type="button"
                  >
                    <span
                      aria-label="arrow-down"
                      class="anticon anticon-arrow-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="arrow-down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M862 465.3h-81c-4.6 0-9 2-12.1 5.5L550 723.1V160c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v563.1L255.1 470.8c-3-3.5-7.4-5.5-12.1-5.5h-81c-6.8 0-10.5 8.1-6 13.2L487.9 861a31.96 31.96 0 0048.3 0L868 478.5c4.5-5.2.8-13.2-6-13.2z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
          style="display: none;"
        />
      </div>
    </main>
  </section>
</DocumentFragment>
`;
